selfdrive/modeld/dmonitoringmodeld.py

```diff
@@ -0,0 +1,160 @@
+
+  def __init__(self):
+    assert ctypes.sizeof(DMonitoringModelResult) == OUTPUT_SIZE * ctypes.sizeof(ctypes.c_float)
+    self.output = np.zeros(OUTPUT_SIZE, dtype=np.float32)
+    self.inputs = {
+      'input_imgs': np.zeros(MODEL_HEIGHT * MODEL_WIDTH, dtype=np.uint8),
+      'calib': np.zeros(CALIB_LEN, dtype=np.float32)}
+
+    self.model = ModelRunner(MODEL_PATHS, self.output, Runtime.DSP, True, None)
+    self.model.addInput("input_imgs", None)
+    self.model.addInput("calib", self.inputs['calib'])
+
+  def run(self, buf:VisionBuf, calib:np.ndarray) -> Tuple[np.ndarray, float]:
+    self.inputs['calib'][:] = calib
+
+    v_offset = buf.height - MODEL_HEIGHT
+    h_offset = (buf.width - MODEL_WIDTH) // 2
+    buf_data = buf.data
+    for row in range(MODEL_HEIGHT):  # make a uint8 copy
+      dst_offset = row * MODEL_WIDTH
+      src_offset = (v_offset + row) * buf.stride + h_offset
+      self.inputs['input_imgs'][dst_offset:dst_offset+MODEL_WIDTH] = buf_data[src_offset:src_offset+MODEL_WIDTH]
+
+    t1 = time.perf_counter()
+    self.model.setInputBuffer("input_imgs", self.inputs['input_imgs'].view(np.float32))
+    self.model.execute()
+    t2 = time.perf_counter()
+    return self.output, t2 - t1
+
+
+def get_driver_state(ds_result: DriverStateResult):
+  return {
+    "faceOrientation": [x * REG_SCALE for x in ds_result.face_orientation],
+    "faceOrientationStd": [math.exp(x) for x in ds_result.face_orientation_std],
+    "facePosition": [x * REG_SCALE for x in ds_result.face_position[:2]],
+    "facePositionStd": [math.exp(x) for x in ds_result.face_position_std[:2]],
+    "faceProb": sigmoid(ds_result.face_prob),
+    "leftEyeProb": sigmoid(ds_result.left_eye_prob),
+    "rightEyeProb": sigmoid(ds_result.right_eye_prob),
+    "leftBlinkProb": sigmoid(ds_result.left_blink_prob),
+    "rightBlinkProb": sigmoid(ds_result.right_blink_prob),
+    "sunglassesProb": sigmoid(ds_result.sunglasses_prob),
+    "occludedProb": sigmoid(ds_result.occluded_prob),
+    "readyProb": [sigmoid(x) for x in ds_result.ready_prob],
+    "notReadyProb": [sigmoid(x) for x in ds_result.not_ready_prob]}
+
+def get_driverstate_packet(model_output: np.ndarray, frame_id: int, location_ts: int, execution_time: float, dsp_execution_time: float):
+  model_result = model_output.ctypes.data_as(ctypes.POINTER(DMonitoringModelResult)).contents
+  msg = messaging.new_message('driverStateV2')
+  msg.driverStateV2 = {
+    'frameId': frame_id,
+    'modelExecutionTime': execution_time,
+    'dspExecutionTime': dsp_execution_time,
+    'poorVisionProb': sigmoid(model_result.poor_vision_prob),
+    'wheelOnRightProb': sigmoid(model_result.wheel_on_right_prob),
+    'leftDriverData': get_driver_state(model_result.driver_state_lhd),
+    'rightDriverData': get_driver_state(model_result.driver_state_rhd),
+    'rawPredictions': model_output.tobytes() if SEND_RAW_PRED else b''}
+
+  return msg
+
+
+def main():
+  gc.disable()
+  set_realtime_priority(53)
```

Comment:
Same as navmodeld, not sure if 53 is the appropriate priority for this process?

Reply:
let's make it 1